src/bgw/job.c

```diff
@@ -176,84 +181,163 @@ ts_bgw_job_get_all(size_t alloc_size, MemoryContext mctx)
+	ts_scanner_foreach(&iterator)
+	{
+		HeapTuple heap = ts_scan_iterator_tuple(&iterator);
+		job = bgw_job_from_tuple(heap, sizeof(BgwJob), mctx);
+		if (!lock_tuple(heap,
+						iterator.ictx.tablerel,
+						tuple_lock_mode,
+						session_lock,
+						&job->tag,
+						block))
+		{
+			/* return NULL if lock could not be acquired */
+			Assert(!block);
+			pfree(job);
+			job = NULL;
+		}
+
+		Assert(num_found == 0);
+		num_found++;
+	}
+
+	return job;
+}
+
+/* Take a share lock on the job tuple ofr the duration of the session. This prevents
+ *  the job from being deleted.
+ *
+ *  Returns whether or not the job still exists.
+ */
+bool
+ts_bgw_job_get_share_lock(int32 bgw_job_id, MemoryContext mctx)
+{
+	/* note the mode here is equivalent to FOR SHARE row locks */
+	BgwJob *job = ts_bgw_job_find_with_lock(bgw_job_id, mctx, RowShareLock, false, true);
+	if (job != NULL)
+	{
+		pfree(job);
+		return true;
+	}
+	return false;
 }
 
 BgwJob *
 ts_bgw_job_find(int32 bgw_job_id, MemoryContext mctx, bool fail_if_not_found)
 {
-	BgwJob *job_stat = NULL;
+	ScanIterator iterator = ts_scan_iterator_create(BGW_JOB, AccessShareLock, mctx);
+	int num_found = 0;
+	BgwJob *job = NULL;
 
-	bgw_job_scan_job_id(bgw_job_id,
-						bgw_job_tuple_found,
-						NULL,
-						&job_stat,
-						mctx,
-						AccessShareLock,
-						fail_if_not_found);
+	init_scan_by_job_id(&iterator, bgw_job_id);
 
-	return job_stat;
+	ts_scanner_foreach(&iterator)
+	{
+		Assert(num_found == 0);
+		job = bgw_job_from_tuple(ts_scan_iterator_tuple(&iterator), sizeof(BgwJob), mctx);
+		num_found++;
+	}
+
+	if (num_found == 0 && fail_if_not_found)
+		elog(ERROR, "job %d not found", bgw_job_id);
+
+	return job;
+	;
 }
 
 static ScanTupleResult
 bgw_job_tuple_delete(TupleInfo *ti, void *data)
 {
 	CatalogSecurityContext sec_ctx;
 	int32 job_id = ((FormData_bgw_job *) GETSTRUCT(ti->tuple))->id;
+	LOCKTAG tag;
+	bool got_lock;
+
+	/* Try getting an exclusive lock on the tuple in a non-blocking manner. Note this is the
+	 * equivalent of a row-based FOR UPDATE lock */
+	got_lock = lock_tuple(ti->tuple,
+						  ti->scanrel,
+						  AccessExclusiveLock,
+						  /* session_lock */ false,
+						  &tag,
+						  /* block */ false);
+	if (!got_lock)
+	{
+		/* If I couldn't get a lock, try killing the background worker that's running the job.
+		 * This is probably not bulletproof but best-effort is good enough here. */
+		VirtualTransactionId *vxid = GetLockConflicts(&tag, AccessExclusiveLock);
+		PGPROC *proc;
+
+		if (VirtualTransactionIdIsValid(*vxid))
+		{
+			proc = BackendIdGetProc(vxid->backendId);
+			if (proc != NULL && proc->isBackgroundWorker)
+			{
+				elog(NOTICE,
+					 "cancelling the background worker for job %d (pid %d)",
+					 job_id,
+					 proc->pid);
+				DirectFunctionCall1(pg_cancel_backend, Int32GetDatum(proc->pid));
+			}
+		}
+
+		/* We have to grab this lock before proceeding so grab it in a blocking manner now */
+		got_lock = lock_tuple(ti->tuple,
```

Comment:
you are not guaranteed to get the lock here if the job has not been cancelled and released all locks. So need to handle the possibility of not getting the lock below?

Reply:
basically, this should be done in a loop...

Reply:
I would go with a finite loop - i.e. attempt to get the lock X number of times, otherwise throw an error and quit.